lib/Analysis/ControlDependenceAnalysis.cpp

```diff
@@ -0,0 +1,298 @@
+        // "least_common_anc" to "block_succ" (including "block_succ") should
+        // be control dependent on "block"
+        // Easy case of block_succ
+        control_deps_map[block_succ].push_back(&block);
+        // Enumerate all paths between "least_common_anc" and "block_succ"
+        SmallVector<PathInDomTree, 4> allPathsFromLeastCommonAncToBlockSucc;
+        enumeratePathsInPostDomTree(least_common_anc, block_succ, &funcReg,
+                                    &postDomTree,
+                                    &allPathsFromLeastCommonAncToBlockSucc);
+
+        for (PathInDomTree path : allPathsFromLeastCommonAncToBlockSucc)
+          for (DominanceInfoNode *domInfo : path) {
+            Block *blockInPath = domInfo->getBlock();
+
+            // Skip the nodes that we have already taken care of above
+            if (blockInPath == least_common_anc || blockInPath == &block ||
+                blockInPath == block_succ)
+              continue;
+
+            // Add block to the control dependencies only if it is not already
+            // there
+            if (std::find(control_deps_map[blockInPath].begin(),
+                          control_deps_map[blockInPath].end(),
+                          &block) == control_deps_map[blockInPath].end())
+              control_deps_map[blockInPath].push_back(&block);
+          }
+      }
+    }
+  }
+
+  // Up to this point, we have correct direct dependencies that do not include
+  // nested dependencies (i.e., dependencies of a block's dependencies) call the
+  // following function to include nested dependencies
+  addDepsOfDeps(funcOp, control_deps_map);
+  all_control_deps_maps.emplace_back(control_deps_map);
+
+  // Extract the forward dependencies for this function
+  identifyForwardControlDeps(funcOp);
+}
+
+void ControlDependenceAnalysis::enumeratePathsInPostDomTree(
+    Block *start_block, Block *end_block, Region *funcReg,
+    llvm::DominatorTreeBase<Block, true> *postDomTree,
+    SmallVector<PathInDomTree, 4> *traversed_nodes) {
+
+  DominanceInfoNode *start_node = postDomTree->getNode(start_block);
+  DominanceInfoNode *end_node = postDomTree->getNode(end_block);
+
+  DenseMap<DominanceInfoNode *, bool> is_visited;
+  for (Block &block : funcReg->getBlocks())
+    is_visited[postDomTree->getNode(&block)] = false;
+
+  PathInDomTree path;
+  path.resize(funcReg->getBlocks().size());
+  int path_index = 0;
+
+  enumeratePathsInPostDomTreeUtil(start_node, end_node, is_visited, path,
+                                  path_index, traversed_nodes);
+}
+
+// DFS to return all nodes in the path between the start_node and end_node (not
+// including start_node and end_node) in the postDom tree
+void ControlDependenceAnalysis::enumeratePathsInPostDomTreeUtil(
+    DominanceInfoNode *start_node, DominanceInfoNode *end_node,
+    DenseMap<DominanceInfoNode *, bool> is_visited, PathInDomTree path,
+    int path_index, SmallVector<PathInDomTree, 4> *traversed_nodes) {
```

Comment:
I supposed `path_index` cannot ever go below 0. If so it should be an `unsigned`.